### PR TITLE
discovery: Add system-probe endpoint for service list

### PR DESCRIPTION
**Pull request by vitkyrka**
### What does this PR do?

Add a new endpoint in system-probe to get the service list with the service names. This will be used from the agent check in a future PR, since the agent check doesn't have permission to get this information.

### Motivation

https://datadoghq.atlassian.net/browse/USMON-1127
https://datadoghq.atlassian.net/browse/USMON-1126
https://datadoghq.atlassian.net/browse/USMON-1128

### CODE REVIEW

File: pkg/process/net/common_linux.go [hunk 1]
```diff
@@ -15,15 +15,16 @@ import (
...
-	netType              = "unix"
+	pingURL               = "http://unix/" + string(sysconfig.PingModule) + "/ping/"
+	tracerouteURL         = "http://unix/" + string(sysconfig.TracerouteModule) + "/traceroute/"
+	connectionsURL        = "http://unix/" + string(sysconfig.NetworkTracerModule) + "/connections"
+	procStatsURL          = "http://unix/" + string(sysconfig.ProcessModule) + "/stats"
+	registerURL           = "http://unix/" + string(sysconfig.NetworkTracerModule) + "/register"
+	statsURL              = "http://unix/debug/stats"
+	pprofURL              = "http://unix/debug/pprof"
+	languageDetectionURL  = "http://unix/" + string(sysconfig.LanguageDetectionModule) + "/detect"
+	discoveryOpenPortsURL = "http://unix/" + string(sysconfig.DiscoveryModule) + "/open_ports"
```
Comment (guyarb) [inline pkg/process/net/common_linux.go:26]: Let's add USM as codeowners for the file

File: pkg/process/net/common.go [hunk 2]
```diff
@@ -379,6 +380,30 @@ func (r *RemoteSysProbeUtil) GetPprof(path string) ([]byte, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("got non-success status code: path %s, url: %s, status_code: %d", r.path, discoveryOpenPortsURL, resp.StatusCode)
+	}
+
+	res := &discoverymodel.OpenPortsResponse{}
```
Comment (guyarb) [inline pkg/process/net/common.go:400]: maybe out of scope, but I do have a concern about the decoder, it can cause memory explosion

File: pkg/collector/corechecks/servicediscovery/model/model.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/model/model.go:4]:
```suggestion
// Copyright 2024-present Datadog, Inc.
```

File: pkg/collector/corechecks/servicediscovery/model/model.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+// Package model contains types for service discovery.
+package model
+
+// Port represents an open port in the current host.
+type Port struct {
+	PID         int    `json:"pid"`
+	ProcessName string `json:"processName"`
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/model/model.go:12]:
```suggestion
	ProcessName string `json:"process_name"`
```

File: pkg/collector/corechecks/servicediscovery/model/model.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package model contains types for service discovery.
+package model
+
+// Port represents an open port in the current host.
+type Port struct {
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/model/model.go:10]: Let's try to find better naming, as port struct that contains port field - feels like a naming issue

File: pkg/collector/corechecks/servicediscovery/model/model.go (hunk 4 above)
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/model/model.go:12]: process name should be out of scope here
we should map to pid, but not to process name

File: pkg/collector/corechecks/servicediscovery/servicediscovery.go [hunk 6]
```diff
@@ -278,6 +279,18 @@ func (c *Check) Interval() time.Duration {
 	return refreshInterval
 }
 
+func portsStr(ports []*model.Port) string {
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/servicediscovery.go:282]: let's document

File: pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go:4]:
```suggestion
// Copyright 2024-present Datadog, Inc.
```

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 8]
```diff
@@ -46,3 +62,30 @@ func (s *discovery) Close() {}
...
+
+func (s *discovery) handleOpenPorts(w http.ResponseWriter, _ *http.Request) {
+	ports, err := s.portPoller.OpenPorts()
+	if err != nil {
+		s.handleError(w, pathOpenPorts, http.StatusInternalServerError, fmt.Errorf("failed to get open ports: %v", err))
+		return
+	}
+
+	var portsResp []*model.Port
+	for _, p := range ports {
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:79]:
```suggestion
	portsResp := make([]*model.Port, len(ports))
	for i, p := range ports {
		portsResp[i] ....
```

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 9]
```diff
@@ -6,24 +6,39 @@
...
+	pathOpenPorts = "/open_ports"
+)
+
 // Ensure discovery implements the module.Module interface.
 var _ module.Module = &discovery{}
 
 // discovery is an implementation of the Module interface for the discovery module.
-type discovery struct{}
+type discovery struct {
+	portPoller *portlist.Poller
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:32]: document

File: pkg/collector/corechecks/servicediscovery/impl_linux.go [hunk 10]
```diff
@@ -87,17 +86,34 @@ func (li *linuxImpl) DiscoverServices() (*discoveredServices, error) {
 		}
 	}
 
-	ports, err := li.portPoller.OpenPorts()
+	sysProbe, err := li.getSysProbeClient()
+	if err != nil {
+		return nil, errWithCode{
+			err:  err,
+			code: errorCodeSystemProbeConn,
+			svc:  nil,
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/impl_linux.go:94]: redundant 
```suggestion
```

File: pkg/collector/corechecks/servicediscovery/impl_linux.go [hunk 11]
```diff
@@ -87,17 +86,34 @@ func (li *linuxImpl) DiscoverServices() (*discoveredServices, error) {
...
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	ports, err := sysProbe.GetDiscoveryOpenPorts(ctx)
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/impl_linux.go:101]: The timeout seems to short.
I think that should be configurable and longer
What is being done in the network check?

File: pkg/collector/corechecks/servicediscovery/impl_linux.go [hunk 12]
```diff
@@ -87,17 +86,34 @@ func (li *linuxImpl) DiscoverServices() (*discoveredServices, error) {
...
+			code: errorCodeSystemProbeOpenPorts,
 			svc:  nil,
 		}
 	}
-	portsByPID := map[int]portlist.List{}
-	for _, p := range ports {
-		portsByPID[p.Pid] = append(portsByPID[p.Pid], p)
+
+	portsByPID := map[int][]*model.Port{}
+	for _, p := range ports.Ports {
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/impl_linux.go:111]: Performance:
```suggestion
	portsByPID := make(map[int][]*model.Port, 0, len(ports.Ports))
	for _, p := range ports.Ports {
```

File: pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+	gorillamux "github.com/gorilla/mux"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/DataDog/datadog-agent/cmd/system-probe/api/module"
+	"github.com/DataDog/datadog-agent/cmd/system-probe/config"
+	"github.com/DataDog/datadog-agent/cmd/system-probe/config/types"
+	workloadmetacomp "github.com/DataDog/datadog-agent/comp/core/workloadmeta/def"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/servicediscovery/model"
+	"github.com/DataDog/datadog-agent/pkg/util/optional"
+	"github.com/stretchr/testify/require"
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go:27]: last import should be moved to the a different import group

File: pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+	}
+	err := module.Register(cfg, mux, []module.Factory{m}, wmeta, nil)
+	require.NoError(t, err)
+
+	srv := httptest.NewServer(mux)
+	t.Cleanup(srv.Close)
+	return srv.URL
+}
+
+func startServerAndGetPort(t *testing.T, modURL string) *model.Port {
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go:57]: what is the intention of the function?

File: pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+	// start a process listening at some port
+	ln, err := net.Listen("tcp", "0.0.0.0:0")
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		_ = ln.Close()
+	})
+	addr := ln.Addr().(*net.TCPAddr)
+
+	req, err := http.NewRequest("GET", modURL+"/discovery/open_ports", nil)
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go:68]: use the consts that we have in the code
also use `http.GetMethod`

**Review by guyarb — CHANGES_REQUESTED** [review]
CRed

File: pkg/process/net/common_linux.go (hunk 1 above)
Comment (hmahmood) [reply pkg/process/net/common_linux.go:26]: Is this endpoint already in system-probe?

Comment (guyarb) [reply pkg/process/net/common_linux.go:26]: it is added in this PR

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 16]
```diff
@@ -46,3 +70,426 @@ func (s *discovery) Close() {}
...
+func (s *discovery) handleGetProc(w http.ResponseWriter, req *http.Request) {
+	vars := mux.Vars(req)
+	pidStr := vars["pid"]
+	pid, err := strconv.ParseUint(pidStr, 10, 32)
+	if err != nil {
+		s.handleError(w, pathGetProc, http.StatusBadRequest, fmt.Errorf("failed to convert pid to integer: %v", err))
+		return
+	}
+
+	if _, err := os.Stat(path.Join(procfs.DefaultMountPoint, pidStr)); os.IsNotExist(err) {
```
Comment (hmahmood) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:139]: The `procfs` path that DD uses can be different from `/proc` and can be set by config. You should use `kernel.HostProc` for this: https://github.com/DataDog/datadog-agent/blob/main/pkg/util/kernel/fs.go#L65

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 17]
```diff
@@ -46,3 +70,426 @@ func (s *discovery) Close() {}
...
+}
+
+func getRootNsInode(procRoot string) (uint32, error) {
+	ns, err := kernel.GetRootNetNamespace(procRoot)
+	if err != nil {
+		return 0, err
+	}
+	inode, err := kernel.GetInoForNs(ns)
+	if err != nil {
+		return 0, err
```
Comment (brycekahle) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:354]: `ns` leaks if this error occurs

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 18]
```diff
@@ -46,3 +70,426 @@ func (s *discovery) Close() {}
...
+	FDs, err := p.OpenFiles()
+	if err != nil {
+		return nil, err
+	}
+
+	// sockets have the following pattern "socket:[inode]"
+	var sockets []uint64
+	for _, fd := range FDs {
+		if strings.HasPrefix(fd.Path, "socket:[") {
+			sock, err := strconv.Atoi(strings.TrimPrefix(fd.Path[:len(fd.Path)-1], "socket:["))
```
Comment (brycekahle) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:183]: consider `strconv.ParseUint` instead of `strconv.Atoi`

Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/module/impl_linux.go:183]: Done.

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go (hunk 17 above)
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/module/impl_linux.go:285]: Fixed.

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go (hunk 16 above)
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/module/impl_linux.go]: I've removed this code in refactoring. I think the other stuff should be fine since gopsutil respects `HOST_PROC` and the agent init sets it if necessary?

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 19]
```diff
@@ -46,3 +70,426 @@ func (s *discovery) Close() {}
...
+			if err != nil {
+				continue
+			}
+			if sock < 0 {
```
[the remark this answers is not in the document]
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/module/impl_linux.go]: The < check is gone after implementing Bryce's suggestion. I've not added the == 0 check, not sure it's really needed? The kernel doesn't use inode 0 for these socket inodes since some years ago but looks like an implementation detail: https://patchwork.kernel.org/project/linux-fsdevel/patch/1435245958-4507-1-git-send-email-cmaiolino@redhat.com/

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 20]
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
...
+	var sockets []uint64
+	for _, fd := range FDs {
+		prefix := "socket:["
+		if strings.HasPrefix(fd.Path, prefix) {
+			inodeStr := strings.TrimPrefix(fd.Path[:len(fd.Path)-1], prefix)
+			sock, err := strconv.ParseUint(inodeStr, 10, 64)
+			if err != nil {
+				continue
+			}
+			sockets = append(sockets, uint64(sock))
```
Comment (brycekahle) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:156]: I think this cast is no longer necessary.

**Review by brycekahle — APPROVED** [review]

File: .github/labeler.yml
Comment (guyarb) [inline .github/labeler.yml:1]: Let's extract it to a different PR

File: .gitlab-ci.yml
Comment (guyarb) [inline .gitlab-ci.yml:1]: Let's extract it to a different PR

File: pkg/collector/corechecks/servicediscovery/model/model.go (hunk 3 above)
Comment (guyarb) [reply pkg/collector/corechecks/servicediscovery/model/model.go:4]: still seeing 2016

File: pkg/collector/corechecks/servicediscovery/impl_linux.go [hunk 21]
```diff
@@ -87,19 +85,24 @@ func (li *linuxImpl) DiscoverServices() (*discoveredServices, error) {
...
+
+	response, err := sysProbe.GetDiscoveryListeners()
+	if err != nil {
+		return nil, errWithCode{
+			err:  err,
+			code: errorCodeSystemProbeOpenPorts,
+		}
 	}
 
+	log.Debugf("listeners: %v", response.Listeners)
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/impl_linux.go:104]: How often is DiscoverServices being called?
if once in 1m, then ignore this comment
if more frequent, then we should wrap this call with a check to the log level, as we've discovered in the past that called `log.Debug` when the log level is info or above, still allocates redundant memory.

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 22]
```diff
@@ -6,24 +6,52 @@
...
+)
+
 // Ensure discovery implements the module.Module interface.
 var _ module.Module = &discovery{}
 
+// cacheData holds process data that should be cached between calls to the
+// endpoint.
+type cacheData struct {
+	env []string
+}
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:42]: Where and How do you ensure the cache cannot bombard the memory of system-probe?

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 23]
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
...
 func (s *discovery) handleStatusEndpoint(w http.ResponseWriter, _ *http.Request) {
 	_, _ = w.Write([]byte("Discovery Module is running"))
 }
+
+func (s *discovery) handleError(w http.ResponseWriter, route string, status int, err error) {
+	_ = log.Errorf("failed to handle /discovery/%s (status: %d): %v", route, status, err)
+	w.WriteHeader(status)
+}
+
+func (s *discovery) handleOpenPorts(w http.ResponseWriter, _ *http.Request) {
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:85]: should we rename the function?

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 24]
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
...
+	"CORECLR_PROFILER_PATH":    {},
+	"CORECLR_ENABLE_PROFILING": {},
+	"JAVA_TOOL_OPTIONS":        {},
+	"_JAVA_OPTIONS":            {},
+	"JDK_JAVA_OPTIONS":         {},
+	"JAVA_OPTIONS":             {},
+	"CATALINA_OPTS":            {},
+	"JDPA_OPTS":                {},
+	"VIRTUAL_ENV":              {},
+}
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:119]: Should be modified to reflect only the environment variables we actually need to save for service-name extraction

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 25]
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
...
+func getSockets(p *process.Process) ([]uint64, error) {
+	FDs, err := p.OpenFiles()
+	if err != nil {
+		return nil, err
+	}
+
+	// sockets have the following pattern "socket:[inode]"
+	var sockets []uint64
+	for _, fd := range FDs {
+		prefix := "socket:["
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:149]: should be declared as a global const to avoid redundant allocations

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 26]
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
...
+	err = kernel.WithNS(ns, func() error {
+		addrs, err := net.InterfaceAddrs()
+		if err != nil {
+			return nil
+		}
+		for _, addr := range addrs {
+			if ipnet, ok := addr.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
+				if ipnet.IP.To4() != nil {
+					out = append(out, ipnet.IP.String())
+				}
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:188]: What about ipv6?

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 27]
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
...
+	// tcpClose is also used to indicate a UDP connection where the other end hasn't been established
+	tcpClose  uint64 = 7
+	udpListen        = tcpClose
+)
+
+// Based on snapshotBoundSockets() in
+// pkg/security/security_profile/activity_tree/process_node_snapshot.go.  Get
+// the list of open ports with socket inodes for the provided namespace.
+func getNsInfo(pid int, isRoot bool) (*namespaceInfo, error) {
+	path := filepath.Join(kernel.HostProc(fmt.Sprintf("%d", pid)))
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:213]: Is `Join` actually relevant here?

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 28]
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
...
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
+	for _, sock := range UDP6 {
+		if sock.St != udpListen {
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:264]: nit: Can that be generalized and reduce redundant duplication?

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 29]
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
...
+}
+
+func (s *discovery) cleanCache(alivePids map[int32]struct{}) {
+	for pid := range s.cache {
+		if _, alive := alivePids[pid]; alive {
+			continue
+		}
+
+		delete(s.cache, pid)
+	}
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:382]: AFAIK, deleting from a map does not free up the memory

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:1]: general comment: Let's add documentation here for the different functions to make it clear for the others

**Comment by guyarb** [comment]
Only a suggestion - I'd split this PR to smaller PRs as it contains many changes, but those are not "atomic" changes.

File: pkg/collector/corechecks/servicediscovery/model/model.go [hunk 30]
```diff
@@ -0,0 +1,20 @@
...
+// Copyright 2016-present Datadog, Inc.
+
+// Package model contains types for service discovery.
+package model
+
+// Listener represents a listening process.
+type Listener struct {
+	PID     int      `json:"pid"`
+	Name    string   `json:"string"`
+	Environ []string `json:"environ"`
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/model/model.go:13]:
```suggestion
	Environ []string `json:"environ,omitempty"`
```

File: pkg/collector/corechecks/servicediscovery/model/model.go [hunk 31]
```diff
@@ -0,0 +1,20 @@
...
+	CWD     string   `json:"cwd"`
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/model/model.go:14]: An improvement - 
Let's treat `/` as the default value, and if it is, then we should not explicitly set it, and by that, sparing more redundant characters

Comment (guyarb) [reply pkg/collector/corechecks/servicediscovery/model/model.go:14]: Why do we need CWD?

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 32]
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
...
+	netNsInfo map[uint32]*namespaceInfo
+}
+
+func (s *discovery) getListener(context parsingContext, pid int32) *model.Listener {
+	proc, err := process.NewProcess(pid)
+	if err != nil {
+		return nil
+	}
+
+	name, err := proc.Name()
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:305]: this is wrong, it returns the process name, instead of the service name

File: pkg/collector/corechecks/servicediscovery/model/model.go [hunk 33]
```diff
@@ -0,0 +1,20 @@
...
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package model contains types for service discovery.
+package model
+
+// Listener represents a listening process.
+type Listener struct {
+	PID     int      `json:"pid"`
+	Name    string   `json:"string"`
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/model/model.go:12]:
```suggestion
	Name    string   `json:"name"`
```

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go (hunk 29 above)
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/module/impl_linux.go:290]: It should free the memory for the strings, right? I added a comment.

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go (hunk 32 above)
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/module/impl_linux.go]: The service name was being calculated in the agent check but now it's moved here.

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go (hunk 22 above)
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/module/impl_linux.go]: Now it's only caching the service name. It's only for running, listening processes so it's proportional to that but I'll create a JIRA to check if we need to limit it in any other way as discussed.

**Comment by vitkyrka** [comment]
> Only a suggestion - I'd split this PR to smaller PRs as it contains many changes, but those are not "atomic" changes.

CI changes split to https://github.com/DataDog/datadog-agent/pull/28193
This PR is to add the endpoint to system-probe.
Next PR will use it in agent check.
Deletion of portlist will be done separately.

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 34]
```diff
@@ -34,15 +63,261 @@ func (s *discovery) GetStats() map[string]interface{} {
...
+				continue
+			}
+			sockets = append(sockets, sock)
+		}
+	}
+	if len(sockets) <= 0 {
+		return nil, nil
+	}
+
+	return sockets, nil
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:125]: nit: shouldn't `return socket, nil` give the same outcome?

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 35]
```diff
@@ -34,15 +63,261 @@ func (s *discovery) GetStats() map[string]interface{} {
...
+	proc, err := process.NewProcess(pid)
+	if err != nil {
+		return nil
+	}
+
+	sockets, err := getSockets(proc)
+	if err != nil {
+		return nil
+	}
+	if sockets == nil {
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:227]: maybe `len(sockets) == 0` will be better

**Review by guyarb — APPROVED** [review]

**Review by hmahmood — APPROVED** [review]

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 36]
```diff
@@ -34,15 +63,261 @@ func (s *discovery) GetStats() map[string]interface{} {
...
+
+// getServiceName gets the service name for a process using the servicedetector
+// module.
+func (s *discovery) getServiceName(proc *process.Process) (string, error) {
+	cmdline, err := proc.CmdlineSlice()
+	if err != nil {
+		return "", nil
+	}
+
+	env, err := proc.Environ()
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/collector/corechecks/servicediscovery/module/impl_linux.go:208]: (can be done in a followup PR)
